### PR TITLE
inductor: use previous guards to know if a size is 1 for broadcasting

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/136640

Today, inductor has some logic to figure out when it needs to do broadcasting during lowering, which just checks if any of the input shapes have sizes equal to 1.

In particular: we should already have this information by the time we get to inductor, because our FakeTensor compute will have branched/guarded on whether any ops performed broadcasting, appropriately.

In particular, if we have a tensor with a size value of `(64//((2048//(s3*((s2//s3)))))))`, and it happens to be equal to one (and it is used in an op that requires this dim to be broadcasted), FakeTensorProp will have generated a guard:
```
Eq((64//((2048//(s3*((s2//s3))))))), 1)
```

I chose the simplest possible way to beef up inductor's checks to know when a given size is equal to 1: loop over the existing shape env guards, and if our current size is a sympy expression on the LHS of one of our `Eq(LHS, 1)` guards, then return True.

I'm hoping for feedback on whether or not this approach is reasonable. One better option I could imagine is that our symbolic reasoning should have automatically simplified the size of our tensor down to a constant as part of evaluating that guard. I was originally going to try to do this directly in the shape env, but I ran into a few issues:

(1) I wanted to call some version of `set_replacement(expr, 1)`. But `set_replacement()` only accepts plain symbols on the LHS, not expressions

(2) in theory I could get this to work if I could rework the above expression to move everything that is not a free variable to the RHS, e.g. `Eq(s2, 32)`. It looks like our existing  `try_solve()` logic is... [not quite able](https://github.com/pytorch/pytorch/blob/main/torch/utils/_sympy/solve.py#L27) to do this generally though.

Checking the guards feels pretty simple-and-easy. Are we worried that it is too slow to iterate over all the guards? I could also cache the lookup so we only need to iterate over guards that are of the form `Eq(LHS, 1)`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138075
* __->__ #136670

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

### CODE REVIEW

**Comment by bdhirsh** [comment]
Updated to just use `shape_env.evaluate_expr(sympy.Eq(expr, 1))` in a bunch of spots relevant to broadcasting in inductor

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -400,9 +400,9 @@ def broadcast_symbolic_shapes(a, b):
     for x, y in itertools.zip_longest(
         reversed(a), reversed(b), fillvalue=sympy.Integer(1)
     ):
-        if y == 1:
+        if V.graph.sizevars.shape_env.evaluate_expr(sympy.Eq(y, 1)):
```
Comment (bdhirsh) [inline torch/_inductor/lowering.py:403]: @ezyang from running my test locally with `TORCH_LOGS="+dynamic"`, this does seem to add extra guards into the shape env. Do you think that's worth worrying about? (or do we have some guard deduping logic later on?)

Comment (ezyang) [reply torch/_inductor/lowering.py:403]: We're supposed to dedupe it! So clean it up into a test case and post a bug about it

**Review by ezyang — APPROVED** [review]
really gotta do that symint rewrite.......

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_inductor/lowering.py:403]: oh nope nvm - the `guard_added` log prints twice but we don't actually end up with duplicate guards in dynamo

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -2271,7 +2271,9 @@ def _normalize_size(x, new_size):
             if new_size[i] == -1:
                 assert old_size[i] is not None
                 new_size[i] = old_size[i]
-            elif old_size[i] is None or old_size[i] == 1:
+            elif old_size[i] is None or V.graph.sizevars.shape_env.evaluate_expr(
```
Comment (bdhirsh) [inline torch/_inductor/ir.py:2274]: welp, I can't actually use evaluate_expr() if the expressions has unbacked symints in it

Comment (bdhirsh) [reply torch/_inductor/ir.py:2274]: ah ok, I think I just want to guard with `size_oblivious=True` (I think this seems... reasonable for cases where inductor is dealing with broadcasting?)

Comment (ezyang) [reply torch/_inductor/ir.py:2274]: Yes, size oblivious is the normal fix here, but you /really/ want to be operating on SymInts here now lol

**Comment by bdhirsh** [comment]
I can repro

**Comment by bdhirsh** [comment]
After staring at the test for a while, I think the test is just flaky and also fails on main (there is a lot of randomness going on in the test). working on an actual repro so I can file a separate issue

File: test/functorch/test_control_flow.py [hunk 3]
```diff
@@ -1562,6 +1562,7 @@ def test_scan_dtype(self, reverse, compile_mode, device, dtype):
     )
     def test_associative_scan_dim(self, combine_mode, reverse, device):
         import random
+        random.seed(10)
```
Comment (bdhirsh) [inline test/functorch/test_control_flow.py:1565]: locally this "fixed" the test for me that broke previously (it was flaky, issue filed here: https://github.com/pytorch/pytorch/issues/137943)

**Comment by ezyang** [comment]
Something like 6% regression on relevant benchmarks

image

**Comment by bdhirsh** [comment]
yep (extra thrash because this PR got reverted and re-landed 😞).

My understanding is that we can either:

(1) accept the regression for now, wait for the `SymInt` re-write in inductor to to claw back the compile time perf (a `SymInt` re-write would mean we automatically get the behavior of not doing sympy logic when our sizes are ints)

(2) if we want to avoid the compile time hit before doing a Symint re-write, we could manually do the branching ourselves in all of the code-paths I changed, although it would be a bit ugly.

**Pull request closed without merging**